### PR TITLE
feat: move API to typescript code generation

### PR DESCRIPTION
**Pull request by xiaozhenliu-gg5**

### CODE REVIEW

**Comment by xiaozhenliu-gg5** [comment]
Can I get some advice of `synth.py` now it has too many post-processing for removing files or replacing file content, is there a faster way to do that? @alexander-fenster Thanks!

File: synth.py [hunk 1]
```diff
@@ -89,8 +82,16 @@
     return new GrafeasClient(this.opts);
   }
 """)
-
 # Node.js specific cleanup
+subprocess.run(['rm', 'src/v1/grafeas_client.ts'])
```
Comment (alexander-fenster) [inline synth.py]: Something like this?

```py
to_remove = []
for root, directories, files in os.walk('.'):
    to_remove.extend([os.path.join(root, file) for file in files if 'grafeas' in file])
for filePath in to_remove:
    os.unlink(filePath)
```

File: synth.py [hunk 2]
```diff
@@ -22,60 +22,53 @@
...
-        'cloud.google.comgrafeas.io',
-        'grafeas.io')
+s.replace('src/v1beta1/*.ts',
+        'grafeas.io',
+        'https://grafeas.io')
 
-# perform surgery inserting the Grafeas client.
-s.replace("src/v1/container_analysis_client.js",
-r"""const path = require\('path'\);
+s.replace('tslint.json', '"extends": "gts/tslint.json"', '"extends": "gts/tslint.json", "linterOptions": {"exclude": ["src/index.ts"]}')
```
Comment (alexander-fenster) [inline synth.py]: Just fix it once and add to `excludes`.

File: synth.py [hunk 3]
```diff
@@ -22,60 +22,53 @@
...
+        )
+    s.copy(library, excludes=['package.json', 'README.md', 'src/index.ts'])
 
 # Copy common templates
 common_templates = gcp.CommonTemplates()
-templates = common_templates.node_library()
+templates = common_templates.node_library(source_location='build/src')
 s.copy(templates)
 
+s.replace('src/v1beta1/index.ts', 'export {GrafeasClient} from \'./grafeas_client\';', '')
```
Comment (alexander-fenster) [inline synth.py]: Just fix it once and add to `excludes`.

File: synth.py [hunk 4]
```diff
@@ -22,60 +22,53 @@
...
+s.replace('src/v1/index.ts', 'export {GrafeasClient} from \'./grafeas_client\';', '')
```
Comment (alexander-fenster) [inline synth.py]: Just fix it once and add to `excludes`.

File: synth.py [hunk 5]
```diff
@@ -22,60 +22,53 @@
...
+s.replace('src/index.ts', 'v1beta1.GrafeasClient', 'v1beta1.GrafeasV1Beta1Client')
```
Comment (alexander-fenster) [inline synth.py]: Just fix it once and add to `excludes`.

File: synth.py [hunk 6]
```diff
@@ -22,60 +22,53 @@
...
+        'grafeas.io',
+        'https://grafeas.io')
 
-# perform surgery inserting the Grafeas client.
-s.replace("src/v1/container_analysis_client.js",
-r"""const path = require\('path'\);
+s.replace('tslint.json', '"extends": "gts/tslint.json"', '"extends": "gts/tslint.json", "linterOptions": {"exclude": ["src/index.ts"]}')
+# perform surgery inserting the Grafeas client. 
+# TODO: replacement not working, ts client further needs opts: ClientOptions
+s.replace("src/v1/container_analysis_client.ts",
```
Comment (alexander-fenster) [inline synth.py]: This "surgery", unfortunately, has to stay, but let's talk tomorrow and see if we can figure out another way of doing this.

**Comment by alexander-fenster** [comment]
OK so removal of grafeas files can be done with a loop as I wrote in comments above. As for the "surgery", I'm thinking of maybe trying a mixin here? Let's talk tomorrow and see if we can make it in a more gentle way.

File: linkinator.config.json [hunk 7]
```diff
@@ -2,7 +2,6 @@
   "recurse": true,
   "skip": [
     "https://codecov.io/gh/googleapis/",
-    "www.googleapis.com",
-    "img.shields.io"
```
Comment (JustinBeckwith) [inline linkinator.config.json]: The PR to include this landed recently in the generator - would be nice to keep this :)

Comment (xiaozhenliu-gg5) [reply linkinator.config.json]: cool, thanks!

File: src/v1/container_analysis_client.ts [hunk 8]
```diff
@@ -253,68 +308,70 @@ class ContainerAnalysisClient {
...
-   *   This object should have the same structure as [Policy]{@link google.iam.v1.Policy}
-   * @param {Object} [options]
-   *   Optional parameters. You can override the default settings for this call, e.g, timeout,
-   *   retries, paginations, etc. See [gax.CallOptions]{@link https://googleapis.github.io/gax-nodejs/interfaces/CallOptions.html} for the details.
-   * @param {function(?Error, ?Object)} [callback]
-   *   The function which will be called with the result of the API call.
-   *
-   *   The second parameter to the callback is an object representing [Policy]{@link google.iam.v1.Policy}.
+   * @param {object} [options]
+   *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
```
Comment (alexander-fenster) [inline src/v1/container_analysis_client.ts]: Interesting: I wonder why we list those extra comments for each field in the request object. That's OK for now, but may I ask you @xiaozhenliu-gg5 to create an issue to figure out if we can have those comments back if it's not too hard? They look useful.

File: src/v1/container_analysis_proto_list.json [hunk 9]
```diff
@@ -1,3 +1,15 @@
...
+  "../../protos/grafeas/v1/cvss.proto",
+  "../../protos/grafeas/v1/deployment.proto",
+  "../../protos/grafeas/v1/discovery.proto",
+  "../../protos/grafeas/v1/grafeas.proto",
+  "../../protos/grafeas/v1/image.proto",
+  "../../protos/grafeas/v1/package.proto",
+  "../../protos/grafeas/v1/provenance.proto",
+  "../../protos/grafeas/v1/upgrade.proto",
+  "../../protos/grafeas/v1/vulnerability.proto"
+]
```
Comment (alexander-fenster) [inline src/v1/container_analysis_proto_list.json]: nit: please add `\n` when you print this file.

File: synth.py [hunk 10]
```diff
@@ -89,8 +97,21 @@
...
-
 # Node.js specific cleanup
+to_remove=['src/v1/grafeas_client.ts', 'src/v1/grafeas_client_config.json', 'src/v1/grafeas_proto_list.json', 'src/v1beta1/grafeas_client.ts','src/v1beta1/grafeas_client_config.json', 'src/v1beta1/grafeas_proto_list.json', 'test/gapic-grafeas_v1_beta1-v1beta1.ts', 'test/gapic-grafeas-v1.ts', 'test/gapic-grafeas-v1beta1.ts']
+for filePath in to_remove:
+    os.unlink(filePath)
+# remove unneeded protos in proto_list.json
+proto_lists=['src/v1/container_analysis_proto_list.json', 'src/v1beta1/container_analysis_v1_beta1_proto_list.json', 'src/v1beta1/grafeas_v1_beta1_proto_list.json']
+remove_proto_keywords=['/google/api', '/google/protobuf', 'google/rpc']
+for file in proto_lists:
+  with open(file, 'r+') as f:
```
Comment (alexander-fenster) [inline synth.py]: you don't need `r+` here, just `r`

File: src/v1/container_analysis_proto_list.json (hunk 9 above)
Comment (xiaozhenliu-gg5) [reply src/v1/container_analysis_proto_list.json]: done, thank you!

File: synth.py (hunk 10 above)
Comment (xiaozhenliu-gg5) [reply synth.py]: yes, you r right, thanks!

File: src/v1/container_analysis_client.ts (hunk 8 above)
Comment (xiaozhenliu-gg5) [reply src/v1/container_analysis_client.ts]: sure, created an issue, will take a look

**Pull request merged**